README.md
# flutter_woocommerce

Let's integrate woo api with flutter easily.

## Getting Started
Import
```dart
import 'package:flutter_woocommerce/flutter_woocommerce.dart';
import 'package:flutter_woocommerce/flutter_woocommerce_utils.dart';
```

Create instance 
```dart
FlutterWoocommerce flutterWoocommerce = FlutterWoocommerce(
      url: <url>,
      consumerKey: <your_consumerKey>,
      consumerSecret: <your_consumerSecret>,
  );
```
## Basic CRUD Requests

### Create
Create data using POST request. Use 
**postReq(** *String* **endPoint**, *Map* **data** )
method. It will return `Json Object` if no error. Then assign it with corresponding Object. And will return **`WooError`** if there is error.   

Example of creating a customer using POST Request
  ```dart
  createCustomer() async {
    Customer newCustomer = Customer(
        firstName: "Ma Tin",
        lastName: "Shwe",
        userName: "matinshwe",
        email: "matinshwe@someone.com");
    newCustomer.password = "12345";
    var result =
        await flutterWoocommerce.postReq('customers', newCustomer.toMap());
    if (result is! WooError) {
      Customer createdCustomer = Customer.fromJSON(result);
      print(createdCustomer.userName + " is created successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }
```


### Read
Read data using GET request. Use **getReq(** *String* **endpoint** **)** method. It will return `Json Object` if no error. Assign it with corresponding Object. And will return **`WooError`** if there is error.   
>*But sometimes you may need to use POST to read data according to condition.*

Example of reading list of customers using GET Request
  ```dart
  fetchAllOrders() async {
    List<Order> orderList = [];
    var result = await flutterWoocommerce.getReq('orders');
    if (result is !WooError) {
      List<dynamic> resultList = result;
      resultList.forEach((order) {
        orderList.add(Order.fromJSON(order));
      });
    }else{
      WooError err = result;
      print(err.message);
    }
    print(orderList.length);
  }
```



### Update
Update data using PUT request. Use **putReq(** *String* **endPoint**, *Map* **data** ) method. It will return `Json Object` if no error. Assign it with corresponding Object. And will return **`WooError`** if there is error.   

Example of updating a product using PUT Request
  ```dart
  updateProduct() async{
    ProductItem productItem = ProductItem(regularPrice: "USD 500");
    var result =
        await flutterWoocommerce.putReq('products/57', productItem.toMap());
    if (result is! WooError) {
      ProductItem createdCustomer = ProductItem.fromJSON(result);
      print(createdCustomer.name + " is updated successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }
```

### Delete
Delete data using DELETE request. Use **deleteReq(** *String* **endPoint**) method. It will return `Json Object` if no error. Assign it with corresponding Object. And will return **`WooError`** if there is error.   

Example of deleting an order using DELETE Request
```dart
  deleteOrder() async {
    var result = await flutterWoocommerce.deleteReq('orders/27');
    if (result is! WooError) {
      Order order = Order.fromJSON(result);
      print(order.id.toString() + " is deleted successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }
```

## Authentication
Intented for customer and admin login. You must install [JWT_auth](https://wordpress.org/plugins/jwt-auth/) plugin first in your wordpress site.

Use **customerLogin(** *User* **user** **)** method to authenticate. It will return **`WooAuthedUser`** object if exist.

For Example:
```dart
  loginCutomer() async{
    var result = await flutterWoocommerce.customerLogin(User(username: 'username', password: "12345"));
    if(result is! WooError){
      WooAuthedUser wooAuthedUser = result;
      print(wooAuthedUser.displayName);
    }else{
      WooError err = result;
      print(err.message);
    }
  }
```

## Bonus Methods
There are some additional methods to save your time.   
```dart
customerSignUp(User user)           => Customer
getCustomers()                      => List<Customer>
updateCustomer(Customer customer)   => Customer
getCustomerByEmail(String email)    => Customer
```

## Objects
Ready to use util objects. Unchecked items are not completed yet.  

- [x] Coupon
- [x] Customer
- [x] Billing
- [x] Shipping
- [x] Order
- [x] LineItem
- [x] Taxes
- [x] TaxLine
- [x] Refund(Order)
- [x] ShippingLine
- [x] FeeLine
- [x] CouponLine
- [ ] ProductItem
- [ ] ProductVar
- [ ] ProductAttr
- [ ] ProductAttrTerms
- [ ] ProductCateg
- [ ] ProductShipClasses
- [ ] ProductTags
- [ ] ProductRev
- [ ] TaxRates
- [ ] TaxClasses
- [ ] Settings
- [ ] SettingsOpt
- [ ] PaymentGatways
- [ ] ShippingZones
- [ ] ShippingZonesMtds
- [ ] ShippingMtds
- [ ] SystemStatus
- [ ] SystemStatTools 
- [ ] data



CHANGELOG.md
## [0.0.3] - 2020-06-11

* fixed security issue

## [0.0.2] - 2020-06-10

* stable version

## [0.0.2-dev.0] - 2020-06-10

* Combined all model files to one file named util.

## [0.0.1] - 2020-06-4

* Documentation added.

## [0.0.1-snowflake.1] - 2020-06-4

* Initial Release.

## [0.0.1-snowflake.0] - 2020-06-4

* Initial Release.

example/lib/main.dart
import 'package:example/config.dart';
import 'package:flutter/material.dart';
import 'package:flutter_woocommerce/flutter_woocommerce.dart';
import 'package:flutter_woocommerce/flutter_woocommerce_utils.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHome(),
    );
  }
}

class MyHome extends StatefulWidget {
  @override
  _MyHomeState createState() => _MyHomeState();
}

class _MyHomeState extends State<MyHome> {
  bool _showLoading = false;

  FlutterWoocommerce flutterWoocommerce = FlutterWoocommerce(
      url: ServerCrends.apiUrl,
      consumerKey: ServerCrends.consumerKey,
      consumerSecret: ServerCrends.consumerSecret);

  createCustomer() async {
    _togggleLoading();
    Customer newCustomer = Customer(
      firstName: "Ma Tin",
      lastName: "Shwe",
      userName: "matinshwe",
    );
    newCustomer.setEmail = "matinshwe@someone.com";
    newCustomer.setPassword = "12345";
    var result =
        await flutterWoocommerce.postReq('customers', newCustomer.toMap());
    _togggleLoading();
    if (result is! WooError) {
      Customer createdCustomer = Customer.fromJSON(result);
      print(createdCustomer.userName + " is created successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }

  fetchAllOrders() async {
    _togggleLoading();
    List<Order> orderList = [];
    var result = await flutterWoocommerce.getReq('orders');
    _togggleLoading();
    if (result is! WooError) {
      List<dynamic> resultList = result;
      resultList.forEach((order) {
        orderList.add(Order.fromJSON(order));
      });

      print(orderList.first.createdVia);
    } else {
      WooError err = result;
      print(err.message);
    }
    print(orderList.length);
  }

  updateProduct() async {
    _togggleLoading();
    ProductItem productItem = ProductItem(regularPrice: "USD 500");
    var result =
        await flutterWoocommerce.putReq('products/57', productItem.toMap());
    _togggleLoading();
    if (result is! WooError) {
      ProductItem createdCustomer = ProductItem.fromJSON(result);
      print(createdCustomer.name + " is updated successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }

  deleteOrder() async {
    _togggleLoading();
    var result = await flutterWoocommerce.deleteReq('orders/27');
    _togggleLoading();
    if (result is! WooError) {
      Order order = Order.fromJSON(result);
      print(order.id.toString() + " is updated successfully!");
    } else {
      WooError err = result;
      print(err.message);
    }
  }

  loginCutomer() async {
    _togggleLoading();
    var result = await flutterWoocommerce
        .customerLogin(User(username: 'matinshwe', password: "12345"));
    _togggleLoading();
    if (result is! WooError) {
      WooAuthedUser wooAuthedUser = result;
      print(wooAuthedUser.displayName);
    } else {
      WooError err = result;
      print(err.message);
    }
  }

  _togggleLoading() {
    if (_showLoading) {
      _showLoading = false;
    } else {
      _showLoading = true;
    }
    setState(() {});
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Demo Flutter Woocommerce'),
        centerTitle: true,
      ),
      body: Container(
        width: double.infinity,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            //There are various way to integrate. This is just example
            Text(
              'See result in console',
              style: TextStyle(color: Colors.black26),
            ),
            SizedBox(
              height: 10,
            ),
            FlatButton(
              padding: EdgeInsets.all(15),
              splashColor: Colors.blue,
              textColor: Colors.blue,
              shape:
                  BeveledRectangleBorder(side: BorderSide(color: Colors.blue)),
              child: _showLoading
                  ? SizedBox(
                      child: CircularProgressIndicator(
                        strokeWidth: 1,
                      ),
                      width: 20,
                      height: 20,
                    )
                  : Text(
                      'Run',
                      style: TextStyle(fontSize: 18),
                    ),
              onPressed: _showLoading
                  ? null
                  : () {
                      fetchAllOrders();
                    },
            ),
          ],
        ),
      ),
    );
  }
}
